### PR TITLE
MAUT-7884 Entity changes for migration for adding unique identifier column

### PR DESCRIPTION
**Pull request by aarohiprasad**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ No]
| New feature/enhancement? (use the a.x branch)      | [ Yes]
| Deprecations?                          | [ NA]
| BC breaks? (use the c.x branch)        | [ NO]
| Automated tests included?              | [ No] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes # [MAUT-7884](https://backlog.acquia.com/browse/MAUT-7884l) 

#### Steps to test this PR:

1. Open this PR for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create a new custom object "Book", try adding fields. A new "is Unique Identifier" toggle button should be visible.
3. Add a name "book uuid" and toggle "is unique identifier to true.
4. Save the custom object.
5. Edit the custom object. Try editing the field "book uuid". The 'is unique identifier' and the 'required' toggle buttons should be readonly.

### CODE REVIEW

**Review by rohitp19 — COMMENTED** [review]
We need to do UI changes in this ticket. Please apply UI changes as discussed.
Thanks

**Review by rohitp19 — COMMENTED** [review]
@aarohiprasad Thanks for adding UI.
There are few suggestions:
1. "Is unique identifier" should be false by default.
2. If property is marked as "Unique" we should mark it as required too and it will be read only so user can not change it to optional.

**Comment by aarohiprasad** [comment]
> @aarohiprasad Thanks for adding UI. There are few suggestions:
> 
> 1. "Is unique identifier" should be false by default.
> 2. If property is marked as "Unique" we should mark it as required too and it will be read only so user can not change it to optional.

@rohitp19 thanks a lot for the review and suggestions! Regarding your concerns


1. This piece of code in Entity/CustomField.php marks it false by default
`  $builder->createField('isUniqueIdentifier', Types::BOOLEAN)
            ->columnName('is_unique_identifier')
            ->option('default', false)
            ->build();`

2. [Form/Type/CustomFieldType.php](https://github.com/acquia/mc-cs-plugin-custom-objects/pull/278/files#diff-e70b4672fa0db517f595e0919ecbae300c66dc53c93a20aebb7b295d7d0ae13d) has code which marks the 'required' and 'is unique identifier' fields as 'readonly' once the field is marked unique. However, I have allowed the following:

- changes in name and other fields
- deletion

Let me know your take on this.

**Review by rohitp19 — COMMENTED** [review]
@aarohiprasad 
We have to make required field validation to readonly only.
Making "Is unique identifier" readonly will be covered by https://backlog.acquia.com/browse/MAUT-7885 ticket which have additional condition there

File: Form/Type/CustomFieldType.php [hunk 1]
```diff
@@ -191,6 +191,7 @@ private function buildModalFormFields(FormBuilderInterface $builder, array $opti
                 'attr'  => [
                     // Need for JS call of method used for changing values dynamically in the frontend. Check co-form.js
                     'data-toggle-button' => true,
+                    'readonly'           => (bool) $customField->getIsUniqueIdentifier(),
```
Comment (dadarya0) [inline Form/Type/CustomFieldType.php:194]: ReadOnly does not work with radio button, readonly is not property of radio button.

File: Assets/js/co-form.js [hunk 2]
```diff
@@ -366,6 +366,14 @@ CustomObjectsForm = {
 
         // Activate content specific stuff
         Mautic.onPageLoad(target, response, true);
+
+        mQuery('#custom_field_isUniqueIdentifier_1').on('change', function() {
+            CustomObjectsForm.setChoiceRequiredVal(true, 'required')
+            mQuery('#objectFieldModal .chosen-required .choice-wrapper').find('label').attr('disabled', true);
```
Comment (dadarya0) [inline Assets/js/co-form.js:372]: Yes/No buttons are still clickable after adding disable attribute in label.

**Review by dadarya0 — CHANGES_REQUESTED** [review]
@aarohiprasad found some minor UI related issue.

File: Form/Type/CustomFieldType.php (hunk 1 above)
Comment (aarohiprasad) [reply Form/Type/CustomFieldType.php:194]: Hi, even I had that doubt. I tried this and this is working okay. This works in the backend by  enabling the "disabled attribute. `<label class="btn btn-default disabled btn-no">`

Comment (dadarya0) [reply Form/Type/CustomFieldType.php:194]: @aarohiprasad  toggle is greyed out and disable so it seems good, no issue.

**Review by dadarya0 — DISMISSED** [review]

**Review by rohitp19 — DISMISSED** [review]
Thanks @aarohiprasad 👍

**Review by dadarya0 — APPROVED** [review]
Thanks

**Comment by aarohiprasad** [comment]
Sorry for that. I found a bug which was causing a 500 error. Fixed that in this commit. Could you please re review? @rohitp19 @dadarya0

**Review by rohitp19 — APPROVED** [review]
Thanks @aarohiprasad 👍

**Comment by escopecz** [comment]
@himanshu8acquia please merge this PR as it went through review and QEs

**Pull request merged**